python/ModelTools.py

```diff
@@ -106,6 +106,9 @@ def __init__(self,datacard,options):
     def setPhysics(self,physicsModel):
         self.physics = physicsModel
         self.physics.setModelBuilder(self)
+        for b in self.DC.bins:
+            for p in self.DC.exp[b].keys():
+                self.physics.tellAboutProcess(b, p)
```

Comment:
It's not clear to me why this is needed - the physics model will already have access to the DC object after L108 here, so if you need to loop through bins and processes in your physics model you can already do this before you reach `getYieldScale`. Or am I missing something?

Reply:
@ajgilbert 

Hello Andrew, 

> About the issue with the RooCategory/flatparameter. 
You can find in these paths, the workspace, log and output: 
/work-zfs/lhc/heshy/anomalouscouplings/scans/cards_fa3fa2fL1fL1Zg_morecategories_finalforthesis/workspace_lumi137.10_scanfa2_fa2,fL1Zg,fa3,fa1,fL1.root
/work-zfs/lhc/heshy/anomalouscouplings/scans/cards_fa3fa2fL1fL1Zg_morecategories_finalforthesis/log_MultiDimFit_obs_lumi137.10_scanfa2_fa2,fL1Zg,fa3,fa1,fL1.txt
/work-zfs/lhc/heshy/anomalouscouplings/scans/cards_fa3fa2fL1fL1Zg_morecategories_finalforthesis/higgsCombine_obs_lumi137.10_scanfa2_fa2,fL1Zg,fa3,fa1,fL1.MultiDimFit.mH125.root
(the RooCategory in question is sgnCMS_zz4l_fai3, which is the sign of fL1).

> About the changes the Physics Model 
Basically we need to determine, before we call getYieldScale, whether or not there are anomalous Hff couplings.  I accomplished that using this function:
https://github.com/cms-analysis/HiggsAnalysis-CombinedLimit/pull/644/files#diff-cb18764af2f0bc5de53451336a743a02148824b812b911c4ba119cc49a994b7eR628-R631
Andrew I can do the modification you suggest here and make a new commit, if you think this is cleaner.

Reply:
Hello Andrew, 

The paths above for the files need to be updated. Let me do this quickly.

Reply:
Here they are 

https://skyriaco.web.cern.ch/skyriaco/public_html/forAndrew/

Reply:
Sorry I don't really understand what I'm supposed to checking in this link. To confirm:

 - You will add a new commit to revert the changes in PhysicsModel.py and ModelTools.py?
 - RooCategory/flatParam: as Nick wrote above

> Now, turning to the behaviour of combine with the RooCategory, what you say about the behaviour of combine is news to me. it's true that combine will do separate fits and then find the one that gives the lowest NLL but for this one has to declare the parameter as a discrete rather than flatParam - at least thats the behaviour I am aware of

I second this, it is not clear to me that flatParam is needed for floating a discrete parameter. Can you explain more clearly why you think it is, ideally with links to the code?

Reply:
In reference to the RooCategory/flatParam thing, there *is* a way that something not declared in the list of discretes can be picked up, by including the runtime definition `ADD_DISCRETE_FALLBACK`. Then the current version of the code will pick up anything with `pdfindex` in the name : https://github.com/cms-analysis/HiggsAnalysis-CombinedLimit/blob/102x/src/Combine.cc#L1226-L1240 

Now the intention was never to then allow the behaviour to pick up any category, but really for debugging certain workspaces (in particular for the diphoton analyses). I think the best would be to just declare `sgnCMS_zz4l_fai3` (which is clearly picked up as a discrete parameter in the output log from @skyriacoCMS) as a `discrete` in the datacard where its needed.

Reply:
> Now the intention was never to then allow the behaviour to pick up any category, but really for debugging certain workspaces (in particular for the diphoton analyses). I think the best would be to just declare `sgnCMS_zz4l_fai3` (which is clearly picked up as a discrete parameter in the output log from @skyriacoCMS) as a `discrete` in the datacard where its needed.

That would be not so desirable for us.  The point is that you can use the same datacard to run for multiple configurations (in particular the order of defining the anomalous couplings).  All couplings get floated, except that the last one is defined as +/-(1 - sum of the others), where only the sign gets floated.  That's decided at the time of text2workspace.  Of course you _can_ write it in the datacard and make a separate datacard for every configuration, but it's ugly.

Reply:
Then why can you not use multiple "sign" categories in the datacard and freeze/float the one that are needed for a specific configuration?  Then you only need one workspace.

Reply:
@ajgilbert 

Hello Andrew, To answer your request on the python files.  
I will make a commit that will change the place of the modifications in the ModelTools.py and add them in the new file SpinZeroStructureAC.py  
I will see what we can do about  PhysicsModel.py. 

Thanks for the feedback.